File: akit/src/main/java/org/littletonrobotics/junction/networktables/LoggedNetworkBoolean.java
// Copyright (c) 2021-2025 Littleton Robotics
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by a BSD
// license that can be found in the LICENSE file
// at the root directory of this project.

package org.littletonrobotics.junction.networktables;

import edu.wpi.first.networktables.BooleanEntry;
import edu.wpi.first.networktables.NetworkTableInstance;
import org.littletonrobotics.junction.LogTable;
import org.littletonrobotics.junction.Logger;
import org.littletonrobotics.junction.inputs.LoggableInputs;

/** Manages a boolean value published to the root table of NT. */
public class LoggedNetworkBoolean extends LoggedNetworkInput {
  private final String key;
  private final BooleanEntry entry;
  private boolean defaultValue = false;
  private boolean value;

  /**
   * Creates a new LoggedNetworkBoolean, for handling a boolean input sent via NetworkTables.
   *
   * @param key The key for the number, published to the root table of NT or
   *     "/DashboardInputs/{key}" when logged.
   */
  public LoggedNetworkBoolean(String key) {
    this.key = key;
    this.entry = NetworkTableInstance.getDefault().getBooleanTopic(key).getEntry(false);
    this.value = defaultValue;
    Logger.registerDashboardInput(this);
  }

  /**
   * Creates a new LoggedNetworkBoolean, for handling a boolean input sent via NetworkTables.
   *
   * @param key The key for the number, published to the root table of NT or
   *     "/DashboardInputs/{key}" when logged.
   * @param defaultValue The default value if no value in NT is found.
   */
  public LoggedNetworkBoolean(String key, boolean defaultValue) {
    this(key);
    setDefault(defaultValue);
    this.value = defaultValue;
  }

  /**
   * Updates the default value, which is used if no value in NT is found.
   *
   * @param defaultValue The new default value.
   */
  public void setDefault(boolean defaultValue) {
    this.defaultValue = defaultValue;
    entry.set(entry.get(defaultValue));
  }

  /**
   * Publishes a new value. Note that the value will not be returned by {@link #get()} until the
   * next cycle.
   *
   * @param value The new value.
   */
  public void set(boolean value) {
    entry.set(value);
  }

  /**
   * Returns the current value.
   *
   * @return The current value.
   */
  public boolean get() {
    return value;
  }

  private final LoggableInputs inputs =
      new LoggableInputs() {
        public void toLog(LogTable table) {
          table.put(removeSlash(key), value);
        }

        public void fromLog(LogTable table) {
          value = table.get(removeSlash(key), defaultValue);
        }
      };

  public void periodic() {
    if (!Logger.hasReplaySource()) {
      value = entry.get(defaultValue);
    }
    Logger.processInputs(prefix, inputs);
  }
}

File: akit/src/main/native/cpp/networktables/LoggedNetworkBoolean.cpp
// Copyright (c) 2021-2026 Littleton Robotics
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by a BSD
// license that can be found in the LICENSE file
// at the root directory of this project.

#include "akit/networktables/LoggedNetworkBoolean.h"
#include "akit/Logger.h"

using namespace akit::nt;

LoggedNetworkBoolean::LoggedNetworkBoolean(std::string key) : key { key }, entry {
		::nt::NetworkTableInstance::GetDefault().GetBooleanTopic(key).GetEntry(
				false) }, value { defaultValue } {
	Logger::RegisterDashboardInput(*this);
}

LoggedNetworkBoolean::LoggedNetworkBoolean(std::string key, bool defaultValue) : LoggedNetworkBoolean {
		key } {
	SetDefault(defaultValue);
	value = defaultValue;
}

void LoggedNetworkBoolean::SetDefault(bool defaultValue) {
	this->defaultValue = defaultValue;
	entry.Set(entry.Get(defaultValue));
}

void LoggedNetworkBoolean::Set(bool value) {
	entry.Set(value);
}

bool LoggedNetworkBoolean::Get() {
	return value;
}

void LoggedNetworkBoolean::ToLog(LogTable &&table) {
	table.Put(RemoveSlash(key), value);
}

void LoggedNetworkBoolean::FromLog(LogTable &&table) {
	value = table.Get(RemoveSlash(key), defaultValue);
}

void LoggedNetworkBoolean::Periodic() {
	if (!Logger::HasReplaySource())
		value = entry.Get(defaultValue);
	Logger::ProcessInputs(std::string { PREFIX }, *this);
}
